packages/alfa-rules/src/sia-r70/rule.ts

```diff
@@ -0,0 +1,138 @@
+import { Page } from "@siteimprove/alfa-web";
+import { List } from "@siteimprove/alfa-list";
+import { Iterable } from "@siteimprove/alfa-iterable";
+import { Array } from "@siteimprove/alfa-array";
+
+import { expectation } from "../common/expectation";
+import { hasChild, isRendered, isDocumentElement } from "../common/predicate";
+
+const { isElement, hasName, hasNamespace } = Element;
+const { and, test } = Predicate;
+
+const isDeprecated = hasName(
+  "acronym",
+  "applet",
+  "basefont",
+  "bgsound",
+  "big",
+  "blink",
+  "center",
+  "content",
+  "dir",
+  "font",
+  "frame",
+  "frameset",
+  "hgroup",
+  "image",
+  "keygen",
+  "marquee",
+  "menuitem",
+  "nobr",
+  "noembed",
+  "noframes",
+  "plaintext",
+  "rb",
+  "rtc",
+  "shadow",
+  "spacer",
+  "strike",
+  "tt",
+  "xmp"
+);
+
+export default Rule.Atomic.of<Page, Document>({
+  uri: "https://alfa.siteimprove.com/rules/sia-r70",
+  evaluate({ device, document }) {
+    return {
+      applicability() {
+        return test(hasChild(isDocumentElement), document) ? [document] : [];
+      },
+
+      expectations(target) {
+        const deprecatedElements = target
+          .descendants({ flattened: true, nested: true })
+          .filter(isElement)
+          .filter(
+            and(hasNamespace(Namespace.HTML), isDeprecated, isRendered(device))
+          );
+
+        return {
+          1: expectation(
+            deprecatedElements.isEmpty(),
+            () => Outcomes.HasNoDeprecatedElement,
+            () => Outcomes.HasDeprecatedElements(deprecatedElements)
+          ),
+        };
+      },
+    };
+  },
+});
+
+export namespace Outcomes {
+  export const HasNoDeprecatedElement = Ok.of(
+    Diagnostic.of(`The document doesn't contain any deprecated element`)
+  );
+
+  export const HasDeprecatedElements = (errors: Iterable<Element>) =>
+    Err.of(
+      DeprecatedElements.of(
+        `The document contains deprecated elements: `,
+        errors
+      )
+    );
+}
+
+class DeprecatedElements extends Diagnostic implements Iterable<Element> {
+  public static of(
+    message: string,
+    errors: Iterable<Element> = []
+  ): DeprecatedElements {
+    return new DeprecatedElements(message, Array.from(errors));
+  }
+
+  private constructor(message: string, errors: Array<Element>) {
+    super(message);
+    this._errors = errors;
+  }
+
+  public *[Symbol.iterator](): Iterator<Element> {
+    yield* this._errors;
+  }
+
+  private _errors: ReadonlyArray<Element>;
+
+  public equalsError(value: unknown, errors: ReadonlyArray<Element>): boolean {
+    return (
+      value instanceof DeprecatedElements &&
+      value._message === this._message &&
+      Array.equals(value._errors, this._errors)
+    );
+  }
+
+  public toJSON(): DeprecatedElements.JSON {
+    return {
+      ...super.toJSON(),
+      errors: Array.toJSON(this._errors),
+    };
+  }
+
+  public get errors(): ReadonlyArray<Element> {
+    return this._errors;
+  }
+}
```

Comment:
```suggestion
class DeprecatedElements extends Diagnostic implements Iterable<Element> {
  public static of(
    message: string,
    errors: Iterable<Element> = []
  ): DeprecatedElements {
    return new DeprecatedElements(message, Array.from(errors));
  }
  
  private _errors: ReadonlyArray<Element>;

  private constructor(message: string, errors: Array<Element>) {
    super(message);
    this._errors = errors;
  }
  
  public get errors(): ReadonlyArray<Element> {
    return this._errors;
  }

  public equals(value: DeprecatedElements): boolean;

  public equals(value: unknown): value is this;

  public equals(value: unknown): boolean {
    return (
      value instanceof DeprecatedElements &&
      value._message === this._message &&
      Array.equals(value._errors, this._errors)
    );
  }
  
  public *[Symbol.iterator](): Iterator<Element> {
    yield* this._errors;
  }

  public toJSON(): DeprecatedElements.JSON {
    return {
      ...super.toJSON(),
      errors: this._errors.map((element) => element.path()),
    };
  }
}
```

Some reorganisation to align with the rest of the classes in Alfa. I also changed the way elements are represented in JSON. Now, they're represented by their path, which is easier for people to grok than their structure.

Reply:
> I also changed the way elements are represented in JSON. Now, they're represented by their path, which is easier for people to grok than their structure.

https://siteimprove.slack.com/archives/G0E9EH2AJ/p1623060925002600
Our main consumer seems to prefer element to their path 😀

Reply:
That relates to the structure of the _class instance_, not its JSON representation. The JSON representation of elements on its own is essentially useless since two structurally equivalent elements sitting different places in the document really aren't the same.